nilearn/_utils/cache_mixin.py

```diff
@@ -108,17 +109,12 @@ def _safe_cache(memory, func, **kwargs):
         with open(version_file, 'r') as _version_file:
             versions = json.load(_version_file)
 
-    modules = (nibabel, )
     # Keep only the major + minor version numbers
-    my_versions = dict((m.__name__, m.__version__) for m in modules)
-    commons = set(versions.keys()).intersection(set(my_versions.keys()))
-    collisions = [
-        m for m in commons if not _compare_version(
-            versions[m], '==', my_versions[m]
-        )
-    ]
-    # Flush cache if version collision
-    if len(collisions) > 0:
+    my_versions = {nibabel.__name__: nibabel.__version__}
```

Comment:
we don't really need to define the `my_versions` collection but could simply compare 
```
    if 'nibabel' in versions and not _nibabel_versions_equal(
            versions['nibabel'], nibabel.__name__
    )
```

Reply:
@jeromedockes we do need `my_versions` near the bottom of `_safe_cache` to overwrite the `version_file` with that nibabel version - also note that you probably meant `nibabel.__version__` instead of `nibabel.__name__` in your snippet. Near the bottom, we see the condition `if versions != my_versions` which doesn't seem to be as thorough as the condition you suggest with `_safe_versions_equal` (using your naming). I think we could safely include overwriting the `version_file` in the same condition such that the patch now looks something like:
```
diff --git a/nilearn/_utils/cache_mixin.py b/nilearn/_utils/cache_mixin.py
index fc8daae3..ff93548a 100644
--- a/nilearn/_utils/cache_mixin.py
+++ b/nilearn/_utils/cache_mixin.py
@@ -80,6 +80,13 @@ def _check_memory(memory, verbose=0):
     return memory


+def _safe_versions_equal(version_a, version_b):
+    try:
+        return _compare_version(version_a, '==', version_b)
+    except TypeError:
+        return False
+
+
 def _safe_cache(memory, func, **kwargs):
     """A wrapper for mem.cache that flushes the cache if the version
     number of nibabel has changed.
@@ -108,17 +115,8 @@ def _safe_cache(memory, func, **kwargs):
         with open(version_file, 'r') as _version_file:
             versions = json.load(_version_file)

-    modules = (nibabel, )
-    # Keep only the major + minor version numbers
-    my_versions = dict((m.__name__, m.__version__) for m in modules)
-    commons = set(versions.keys()).intersection(set(my_versions.keys()))
-    collisions = [
-        m for m in commons if not _compare_version(
-            versions[m], '==', my_versions[m]
-        )
-    ]
-    # Flush cache if version collision
-    if len(collisions) > 0:
+    my_versions = {nibabel.__name__: nibabel.__version__}
+    if not _safe_versions_equal(versions.get('nibabel'), my_versions.get('nibabel')):
         if nilearn.CHECK_CACHE_VERSION:
             warnings.warn("Incompatible cache in %s: "
                           "different version of nibabel. Deleting "
@@ -146,8 +144,6 @@ def _safe_cache(memory, func, **kwargs):
             warnings.warn("Incompatible cache in %s: "
                           "old version of nibabel." % location)

-    # Write json files if configuration is different
-    if versions != my_versions:
         with open(version_file, 'w') as _version_file:
             json.dump(my_versions, _version_file)

```

Reply:
Note that the above changes still work given a few tests like:
```
diff --git a/nilearn/tests/test_cache_mixin.py b/nilearn/tests/test_cache_mixin.py
index 7687b37e..c6b8989f 100644
--- a/nilearn/tests/test_cache_mixin.py
+++ b/nilearn/tests/test_cache_mixin.py
@@ -10,11 +10,15 @@ from pathlib import Path
 import pytest

 from joblib import Memory
+from nibabel._version import get_versions

 import nilearn
 from nilearn._utils import cache_mixin, CacheMixin


+_NIBABEL_VERSION = get_versions()['version']
+
+
 def _get_subdirs(top_dir):
     top_dir = Path(top_dir)
     children = list(top_dir.glob("*"))
@@ -182,3 +186,23 @@ def test_cache_shelving():
         res = cache_mixin.cache(f, mem, shelve=True)(2)
         assert res.get() == 2
         assert len(_get_subdirs(joblib_dir)) == 1
+
+
+@pytest.mark.parametrize('old_versions,new_versions', [
+    ({}, {'nibabel': _NIBABEL_VERSION}),
+    ({'nibabel': '0.0.0'}, {'nibabel': _NIBABEL_VERSION}),
+    ({'nibabel': '1000.0.0'}, {'nibabel': _NIBABEL_VERSION}),
+    ({'nibabel': 0}, {'nibabel': _NIBABEL_VERSION}),
+    ({'test': _NIBABEL_VERSION}, {'nibabel': _NIBABEL_VERSION}),
+])
+def test_safe_cache(old_versions, new_versions):
+    with tempfile.TemporaryDirectory() as temp_dir:
+        version_file = Path(temp_dir) / 'joblib' / 'module_versions.json'
+        version_file.parent.mkdir()
+        version_file.write_text(json.dumps(old_versions))
+
+        memory = Memory(location=temp_dir)
+        cache_mixin._safe_cache(memory, f)
+
+        versions = json.loads(version_file.read_text())
+        assert versions == new_versions
```